### PR TITLE
feat: add dashboard visualization support to extension SDK

### PR DESCRIPTION
**Pull request by bryans99**
This is a work in progress.

Tests and types need to be added.

The playground changes may be reverted and a playground-viz maybe created instead. 

The PR itself may be abandoned in favor of another. Putting this up to allow Will to test if he wants to.

### CODE REVIEW

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:118]: do we use this convention in multiple places? would it be worth creating a trivial function for console errors instead and decorating the lint override in one place?

File: packages/extension-sdk/src/visualization/types.ts
Comment (jkaster) [inline packages/extension-sdk/src/visualization/types.ts:31]: Looking forward to seeing this fleshed out

File: packages/extension-tile-playground/README.md
Comment (jkaster) [inline packages/extension-tile-playground/README.md:6]:
```suggestion
quickly prototyped and tested in dashboard mount points. Do not save prototyping to this repo as this
```

**Review by jkaster — COMMENTED** [review]
Looks good so far. Did you want me to try the branch or something?

File: packages/extension-sdk/src/visualization/types.ts
Comment (bryans99) [reply packages/extension-sdk/src/visualization/types.ts]: Adapted from here - https://github.com/looker-open-source/custom_visualizations_v2/blob/master/src/examples/types/types.ts

It's a start

File: packages/extension-tile-playground/README.md
Comment (bryans99) [reply packages/extension-tile-playground/README.md]: Fixed

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:67]: Bit by English English vs. American English confusion here, I bet.

```suggestion
      visualizationDataReceivedCallback,
```

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:72]:
```suggestion
    this.visualizationDataReceivedCallback = visualizationDataReceivedCallback
```

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:96]: here too but I'm sure you'll catch it when you rename refactor

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:381]: this can be removed

File: packages/extension-sdk/src/util/logger.ts
Comment (jkaster) [inline packages/extension-sdk/src/util/logger.ts:28]: thanks for humoring me

File: packages/extension-sdk/src/visualization/types.ts
Comment (jkaster) [inline packages/extension-sdk/src/visualization/types.ts:30]:
```suggestion
export type VisualizationDataReceivedCallback = (
```

File: packages/extension-tile-playground/README.md
Comment (jkaster) [inline packages/extension-tile-playground/README.md:69]: Didn't you create a `yarn dev:play` alias for this?

**Review by jkaster — CHANGES_REQUESTED** [review]
Minor typos

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts]: Dammit! I keep wanting to type viz! Fixed

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts]: Fixed

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts]: Doh! removed!

File: packages/extension-tile-playground/README.md
Comment (bryans99) [reply packages/extension-tile-playground/README.md:69]: I did. A changed for this will be in the next commit

File: packages/extension-playground/package.json
Comment (jkaster) [inline packages/extension-playground/package.json:15]: is this change intentional?

File: packages/extension-tile-playground/package.json
Comment (jkaster) [inline packages/extension-tile-playground/package.json:15]: copy/pasta for extension-api-explorer?

**Review by jkaster — COMMENTED** [review]
Might have some copy/pasta issues?

**Review by jerk1022 — APPROVED** [review]

File: packages/extension-sdk/src/util/index.ts
Comment (whscullin) [inline packages/extension-sdk/src/util/index.ts:5]: Are we automatically generating the wrong year? Or does this placate the linter?

File: packages/extension-playground/package.json
Comment (bryans99) [reply packages/extension-playground/package.json]: Nope. Copy/pasta. Removing watch as not needed.

File: packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx
Comment (jkaster) [inline packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx:70]: when do you think this will switch to `core40SDK`? Do we need a deprecation schedule for that?

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:83]: when would this return undefined?

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:90]: when would this return undefined?

File: packages/extension-sdk/src/connect/tile/types.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/tile/types.ts:26]: PIVOT!

**Review by jkaster — COMMENTED** [review]
Still a draft, so I won't approve yet but only minor comments from me

File: packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx
Comment (bryans99) [reply packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx:70]: Yeah, we should probably deprecate ExtensionProvider in favor of ExtensionProvider2. What's the process for doing that in open source? Do we need google approval? Do we just mark it as deprecated?

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts:83]: It would not. A left over from the original implementation when I only created it if visualizationDataReceivedCallback was set. Will fix

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts:90]: Copy pasta. Will be fixed

File: packages/extension-sdk/src/connect/tile/types.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/tile/types.ts:63]: ????

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.ts:131]: Is `tileHostDataChanged` a sub-event of `tileHostChangedCallback`?

**Review by jkaster — COMMENTED** [review]
Looking good. Minor question.

File: packages/extension-sdk/src/connect/tile/types.ts
Comment (jkaster) [reply packages/extension-sdk/src/connect/tile/types.ts:63]: Lame joke. https://www.youtube.com/watch?v=n67RYI_0sc0

File: packages/extension-sdk/src/connect/extension_host_api.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts:131]: Good catch. I renamed it and I guess I missed that one. Another change coming

Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.ts]: Done

**Review by jkaster — COMMENTED** [review]
Updates look good

**Review by jkaster — CHANGES_REQUESTED** [review]

File: packages/extension-sdk-react/src/components/ExtensionProvider40/ExtensionProvider40.spec.tsx
Comment (bryans99) [reply packages/extension-sdk-react/src/components/ExtensionProvider40/ExtensionProvider40.spec.tsx:49]: fixed

Comment (bryans99) [reply packages/extension-sdk-react/src/components/ExtensionProvider40/ExtensionProvider40.spec.tsx:49]: Actually, I thought I was looking at another PR so not resolved. I broke this change out into a separate PR. That PR will hopefully be merged with this and this particular issue will become moot.

Comment (jkaster) [reply packages/extension-sdk-react/src/components/ExtensionProvider40/ExtensionProvider40.spec.tsx:49]: Yeah, I noticed the duplicate

File: packages/extension-sdk/src/connect/extension_host_api.spec.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/extension_host_api.spec.ts:177]: that's an old version!

**Review by jkaster — APPROVED** [review]
Didn't notice anything on this pass

File: packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx
Comment (bryans99) [reply packages/extension-sdk-react/src/components/ExtensionProvider/ExtensionProvider.tsx:74]: I've created ExtensionProvider40 in a separate PR. When that it published I will update all of the examples to use it. After that we can mark ExtensionProvider and ExtensionProvider2 as deprecated. Not what the process is. I can update the release notes for extension sdk react to indicate that ExtensionProvider40 is the preferred provider and that the other two will be deprecated but supported for at least a year. Or we can keep ExtenionProvider but mark coreSDK31 and coreSDK properties as deprecated. c

File: packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts:97]: I love Tigger as much as the next guy, but maybe you mean 

```suggestion
  it('sends trigger message ', () => {
```

File: packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts:109]: Is this a duplicate test?
```suggestion
  it('sends trigger message ', () => {
```

Comment (bryans99) [reply packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts:109]: Yes - will delete

File: packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/tile/tile_sdk.spec.ts:97]: Will fix

File: packages/extension-sdk/src/connect/extension_host_api.spec.ts
Comment (bryans99) [reply packages/extension-sdk/src/connect/extension_host_api.spec.ts]: Updated to a more recent version

File: packages/extension-sdk-react/src/components/utils/get_initial_route_entries.ts
Comment (jkaster) [inline packages/extension-sdk-react/src/components/utils/get_initial_route_entries.ts:42]: BTW, to force a publish you could just use a 

```ts
/** temporary comment to change this package signature /*
```

header comment somewhere

File: packages/extension-sdk/src/connect/tile/tile_sdk.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/tile/tile_sdk.ts:59]: seems to be a repeated pattern in this file. Does it make any sense to try to abstract?

File: packages/extension-sdk/src/connect/visualization/visualization_sdk.ts
Comment (jkaster) [inline packages/extension-sdk/src/connect/visualization/visualization_sdk.ts:27]: two `NOT` s?

File: packages/extension-sdk/src/util/errors.ts
Comment (jkaster) [inline packages/extension-sdk/src/util/errors.ts:26]:
```suggestion
export const DASHBOARD_MOUNT_NOT_SUPPORTED_ERROR = new Error(
```

**Review by jkaster — APPROVED** [review]
Looks workable but you may want to take a suggested rename

**Comment by bryans99** [comment]
This branch has been merged into the `feature/extension-dashboard-tile` branch which is the official WIP branch for this feature

**Pull request closed without merging**
